File: 216-combination-sum-iii/216-combination-sum-iii.cpp
class Solution {
public:
    void SumUntil(vector<vector<int>>& ans, vector<int>& combination, int k, int target, int start){
        if(combination.size() == k && target == 0)
        {
            ans.push_back(combination);
            return;
        }
        // if(combination.size() < k)
    
          for(int i = start; i <= 9; i++)
          {
                if(target - i < 0)
                    break;
                combination.push_back(i);
                SumUntil(ans, combination, k, target-i,i+1);
                combination.pop_back();
            }  
        
    }
    
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> ans;
        vector<int> combination;

        SumUntil(ans,combination,k,n,1);
        return ans;
    }
};

File: 216-combination-sum-iii/216-combination-sum-iii.java
class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> ans = new ArrayList<>();
        sumUntil(ans, new ArrayList<Integer>(), k, n, 1);
        return ans;
    }
    
    public void sumUntil(List<List<Integer>> ans, List<Integer> comb, int k, int target, int start)
    {
        if(comb.size() == k && target == 0)
        {
            List<Integer> li = new ArrayList<Integer>(comb);
            ans.add(li);
            return;
        }
        
        if(comb.size() < k)
        {
            for(int i = start; i <= 9; i++)
            {
                if(target-i < 0)
                    break;
                comb.add(i);
                sumUntil(ans, comb, k, target-i, i+1);
                comb.remove(comb.size()-1);
            }
        }
    }
}